### PR TITLE
Rebuild Coursera branch off master

### LINKED ISSUE #262: Port changes from master to coursera
I have just discovered that we did not have an issue for this thing we've been working on for a while now, so I'll create one to document progress.

- [ ] Week 1 (#235, #236, #277, #278, #280; `primer/` is still missing altogether)
- [x] Week 2 (finished in #255)
- [ ] Week 3
- [ ] Week 4
- [ ] Week 5 (#243 and #245, but that only covers 1/2 notebooks)
- [ ] Week 6 (#292)

Also:

* #259, #260 sync metadata changes (by getting rid of it entirely)
* #261 ports changes to all scripts in a very direct way

Here is a convenient CSV that matches files in `master` and `coursera`:

```csv
week01_intro/crossentropy_method.ipynb,week1_intro/crossentropy_method.ipynb
week01_intro/deep_crossentropy_method.ipynb,week1_intro/deep_crossentropy_method.ipynb
week01_intro/pong.py,
week01_intro/primer_python_for_ml/recap_ml.ipynb,week1_intro/primer/recap_ml.ipynb
week01_intro/primer_python_for_ml/train.csv,week1_intro/primer/train.csv
week01_intro/project_starter_evolution_strategies.ipynb,
week01_intro/seminar_gym_interface.ipynb,week1_intro/gym_interface.ipynb
week02_value_based/mdp.py,week2_model_based/mdp.py
week02_value_based/seminar_vi.ipynb,week2_model_based/practice_vi.ipynb
week03_model_free/homework.ipynb,week3_model_free/sarsa.ipynb;week3_model_free/experience_replay.ipynb
week03_model_free/seminar_qlearning.ipynb,week3_model_free/qlearning.ipynb
week04_[recap]_deep_learning/fix_my_nn.ipynb,
week04_[recap]_deep_learning/mnist.py,week1_intro/primer/mnist.py
week04_[recap]_deep_learning/notmnist.py,
week04_[recap]_deep_learning/practice_lasagne.ipynb,
week04_[recap]_deep_learning/seminar_tensorflow.ipynb,week1_intro/primer/recap_tensorflow.ipynb
week04_[recap]_deep_learning/seminar_pytorch.ipynb,week1_intro/primer/recap_pytorch.ipynb
week04_[recap]_deep_learning/seminar_tensorflow.ipynb,
week04_approx_rl/atari_wrappers.py,week4_approx/atari_wrappers.py
week04_approx_rl/framebuffer.py,week4_approx/framebuffer.py
week04_approx_rl/homework_lasagne.ipynb,
week04_approx_rl/homework_pytorch_debug.ipynb,
week04_approx_rl/homework_pytorch_main.ipynb,week4_approx/dqn_atari_pytorch.ipynb
week04_approx_rl/homework_tf.ipynb,week4_approx/dqn_atari.ipynb
week04_approx_rl/replay_buffer.py,week4_approx/replay_buffer.py
week04_approx_rl/seminar_lasagne.ipynb,
week04_approx_rl/seminar_pytorch.ipynb,week4_approx/practice_approx_qlearning_pytorch.ipynb
week04_approx_rl/seminar_tf.ipynb,week4_approx/practice_approx_qlearning.ipynb
week04_approx_rl/utils.py,week4_approx/utils.py
week05_explore/bayes.py,
week05_explore/week5.ipynb,week6_outro/bandits.ipynb
week06_policy_based/a2c-optional.ipynb,
week06_policy_based/atari_wrappers.py,
week06_policy_based/env_batch.py,
week06_policy_based/reinforce_lasagne.ipynb,
week06_policy_based/reinforce_pytorch.ipynb,
week06_policy_based/reinforce_tensorflow.ipynb,week5_policy_based/practice_reinforce.ipynb
week06_policy_based/runners.py,
week07_[recap]_rnn/seminar_lasagne.ipynb,
week07_[recap]_rnn/seminar_lasagne_ingraph.ipynb,
week07_[recap]_rnn/seminar_pytorch.ipynb,
week07_[recap]_rnn/seminar_tf.ipynb,
week07_seq2seq/basic_model_tf.py,week6_outro/seq2seq/basic_model_tf.py
week07_seq2seq/basic_model_theano.py,
week07_seq2seq/basic_model_torch.py,
week07_seq2seq/bonus.ipynb,
week07_seq2seq/practice_tf.ipynb,week6_outro/seq2seq/practice_tf.ipynb
week07_seq2seq/practice_theano.ipynb,
week07_seq2seq/practice_torch.ipynb,
week07_seq2seq/voc.py,week6_outro/seq2seq/voc.py
week08_pomdp/atari_util.py,week5_policy_based/atari_util.py
week08_pomdp/env_pool.py,
week08_pomdp/homework_common_part2.ipynb,
week08_pomdp/practice_pytorch.ipynb,
week08_pomdp/practice_tensorflow.ipynb,week5_policy_based/practice_a3c.ipynb
week08_pomdp/practice_theano.ipynb,
week08_pomdp/theano_optional_recurrence_tutorial.ipynb,
week09_policy_II/mujoco_wrappers.py,
week09_policy_II/ppo.ipynb,
week09_policy_II/runners.py,
week09_policy_II/seminar_TRPO_pytorch.ipynb,
week09_policy_II/seminar_TRPO_tensorflow.ipynb,
week09_policy_II/seminar_TRPO_theano.ipynb,
week10_planning/seminar_MCTS.ipynb,week6_outro/practice_mcts.ipynb
```

I recommend to alias `column -s, -t` as `csv` in your shell to display CSVs.

### PR DESCRIPTION
**Pull request by dniku**
Rebuild branch off the most recent `spring20` (actually, `de86c56`). Here is how I did it.

I checked out the repository into multiple working trees: `master`, `coursera`, and `coursera-merge`:

```bash
git checkout spring20  # current substitute for master
git worktree add ../coursera coursera
git branch coursera-merge coursera
git worktree add ../coursera-merge coursera-merge
```

Then I wrote the following script:

```python
import shutil
from pathlib import Path

import pandas as pd

master_root = Path('master')
coursera_old_root = Path('coursera')
coursera_new_root = Path('coursera-merge')

master_coursera_matching = pd.read_csv(master_root / 'master_coursera_matching.csv', header=None, names=['master', 'coursera'])

master_ignored = {
    # Needs to be split
    'week03_model_free/homework.ipynb',
    # PyTorch
    'week04_[recap]_deep_learning/seminar_pytorch.ipynb',
    'week04_approx_rl/atari_wrappers.py',
    'week04_approx_rl/homework_pytorch_main.ipynb',
    'week04_approx_rl/seminar_pytorch.ipynb',
    'week04_approx_rl/utils.py',
    # These do not exactly correspond to Week 5 assignment
    'week08_pomdp/atari_util.py',
    'week08_pomdp/practice_tensorflow.ipynb',
}

files_to_copy = {
    coursera_old_root / path : coursera_new_root / path
    for path in [
        'week1_intro/submit.py',
        'week2_model_based/submit.py',
        'week3_model_free/submit.py',
        'week4_approx/submit.py',
        'week5_policy_based/submit.py',
        'week6_outro/seq2seq/he-pron-wiktionary.txt',
        'week6_outro/seq2seq/main_dataset.txt',
        'week6_outro/seq2seq/submit.py',
        'week6_outro/submit.py',
        '.gitignore',
        'grading.py',
        'xvfb'
    ]
}

for row in master_coursera_matching.itertuples():
    if isinstance(row.coursera, float):
        continue
    if row.master in master_ignored:
        continue
    master = master_root / row.master
    coursera = coursera_old_root / row.coursera
    coursera_new = coursera_new_root / row.coursera
    assert master.exists(), master
    assert coursera.exists(), coursera
    files_to_copy[master] = coursera_new

for src, dst in files_to_copy.items():
    Path(dst.parent).mkdir(parents=True, exist_ok=True)
    shutil.copyfile(src, dst)
```

and placed it in `../build_coursera.py`. `master_coursera_matching.csv` is posted in #262. Then I did:

```bash
cd ..
rm -r coursera-merge/*
python build_coursera.py
```

This copied the most recent versions of all notebooks from `spring20` to `coursera-merge`. After that, I went through all notebooks with `git cola` and `nbdiff` and reverted changes that are irrelevant to Coursera (primarily those about points and submission cells).

TODO:

- [x] Either split `week03_model_free/homework.ipynb` on `spring20` or transfer parts of it manually into `week3_model_free/sarsa.ipynb` and `week3_model_free/experience_replay.ipynb`.
- [x] Finish porting Week 5 (`practice_a3c.ipynb` specifically). Actually, it seems that the port should be in the opposite direction, as some bugs seem to have been fixed in `coursera`, but not in `master`. Not sure how that is possible since I had been certain we had finished porting `coursera` to `master`.
- [x] Review everything _very_ thoroughly. There are a lot of changes.

When this PR is completed, it should fix #262 and #173.

### CODE REVIEW

File: week6_outro/practice_mcts.ipynb [hunk 1]
```diff
@@ -411,22 +436,21 @@
...
+    "        node = <select best leaf >\n",
     "\n",
     "        if node.is_done:\n",
     "            node.propagate(0)\n",
     "\n",
-    "        else: #node is not terminal\n",
-    "            <expand-simulate-propagate loop>\n",
-    "    \n"
+    "        else:  # node is not terminal\n",
+    "            <expand-simulate-propagate loop >"
```
Comment (dniku) [inline week6_outro/practice_mcts.ipynb:453]: Should be overwritten

File: week6_outro/seq2seq/practice_tf.ipynb [hunk 2]
```diff
@@ -10,7 +10,7 @@
     "\n",
     " * word (sequence of letters in source language) -> translation (sequence of letters in target language)\n",
     "\n",
-    "Unlike most deep learning researchers do, we won't only train it to maximize likelihood of correct translation, but also employ reinforcement learning to actually teach it to translate with as few errors as possible.\n",
+    "Unlike what most deep learning practicioners do, we won't only train it to maximize likelihood of correct translation, but also employ reinforcement learning to actually teach it to translate with as few errors as possible.\n",
```
Comment (dniku) [inline week6_outro/seq2seq/practice_tf.ipynb:13]: Typo, will be overwritten

**Comment by dniku** [comment]
Except for `week5/practice_a3c.ipynb`, the whole thing is ready for review.

File: week3_model_free/experience_replay.ipynb [hunk 3]
```diff
@@ -156,9 +190,7 @@
    "outputs": [],
    "source": [
     "import gym\n",
-    "from qlearning import QLearningAgent\n",
```
Comment (yhn112) [inline week3_model_free/experience_replay.ipynb:159]: Please, check if we still need qlearning.py in repo

File: week4_approx/dqn_atari.ipynb [hunk 4]
```diff
@@ -47,7 +47,7 @@
    "metadata": {},
    "source": [
     "### Let's play some old videogames\n",
-    "![img](https://s17.postimg.cc/y9xcab74f/nerd.png)\n",
+    "![img](https://github.com/yandexdataschool/Practical_RL/raw/master/yet_another_week/_resource/nerd.png)\n",
```
Comment (yhn112) [inline week4_approx/dqn_atari.ipynb:50]: Don't we need to host images somewhere else?

File: week4_approx/dqn_atari.ipynb [hunk 5]
```diff
@@ -75,29 +75,30 @@
...
     "        \"\"\"what happens to each observation\"\"\"\n",
-    "        \n",
+    "\n",
     "        # Here's what you need to do:\n",
     "        #  * crop image, remove irrelevant parts\n",
-    "        #  * resize image to self.img_size \n",
-    "        #     (use imresize imported above or any library you want,\n",
-    "        #      e.g. opencv, skimage, PIL, keras)\n",
+    "        #  * resize image to self.img_size\n",
+    "        #     (use resize imported above or any library you want,\n",
```
Comment (yhn112) [inline week4_approx/dqn_atari.ipynb:95]: Let's rewrite this part to opencv

File: week4_approx/practice_approx_qlearning.ipynb [hunk 6]
```diff
@@ -99,8 +99,7 @@
     "network = keras.models.Sequential()\n",
     "network.add(L.InputLayer(state_dim))\n",
     "\n",
-    "# let's create a network for approximate q-learning following guidelines above\n",
```
Comment (yhn112) [inline week4_approx/practice_approx_qlearning.ipynb:102]: Why not?

File: week5_policy_based/practice_reinforce.ipynb [hunk 7]
```diff
@@ -15,48 +15,28 @@
...
     "    os.environ['DISPLAY'] = ':1'"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Some caveats:\n",
+    "\n",
+    "* The notebook assumes that you have [OpenAI Gym](https://github.com/openai/gym) installed.\n",
```
Comment (yhn112) [inline week5_policy_based/practice_reinforce.ipynb:30]: Week 5? We need gym? Really?

File: week6_outro/practice_mcts.ipynb [hunk 8]
```diff
@@ -203,13 +213,30 @@
...
+    }
+   },
    "cell_type": "markdown",
    "metadata": {},
    "source": [
     "# MCTS: Monte-Carlo tree search\n",
     "\n",
+    "![MCTS.png](attachment:MCTS.png)\n",
```
Comment (yhn112) [inline week6_outro/practice_mcts.ipynb:226]: Let's choose one of:
1) embed images to notebooks
2) host images in our repo
3) host images on some image hosting
4) find all the images on the existing sources.
Currently we use all 4 ways in different places.

File: week3_model_free/experience_replay.ipynb (hunk 3 above)
Comment (dniku) [reply week3_model_free/experience_replay.ipynb:159]: Removed `qlearning.py` from the repo.

File: week4_approx/dqn_atari.ipynb (hunk 4 above)
Comment (dniku) [reply week4_approx/dqn_atari.ipynb:50]: We do, but that's out of scope for this PR.

File: week6_outro/practice_mcts.ipynb (hunk 8 above)
Comment (dniku) [reply week6_outro/practice_mcts.ipynb:226]: Out of scope for this PR, but certainly worth creating an issue

File: README.md [hunk 9]
```diff
@@ -1,134 +1,3 @@
...
-  * Lecture: Trust region policy optimization. NPO/PPO. Deterministic policy gradient. DDPG. Bonus: DPG for discrete action spaces.
-  * Seminar: Approximate TRPO for simple robotic tasks.
-
-[Maybe a bonus lecture here]
-
-
-
-
-
-# Course staff
```
[the remark this answers is not in the document]
Comment (dniku) [reply README.md:115]: Too hard to maintain two copies. We can reconsider later

**Pull request merged**